File: app/src/main/java/com/kazakago/cleanarchitecture/data/mapper/WeatherMapper.java
package com.kazakago.cleanarchitecture.data.mapper;

import android.support.annotation.Nullable;

import com.kazakago.cleanarchitecture.data.entity.weather.WeatherEntity;
import com.kazakago.cleanarchitecture.domain.model.weather.CopyrightModel;
import com.kazakago.cleanarchitecture.domain.model.weather.DescriptionModel;
import com.kazakago.cleanarchitecture.domain.model.weather.ForecastModel;
import com.kazakago.cleanarchitecture.domain.model.weather.ImageModel;
import com.kazakago.cleanarchitecture.domain.model.weather.LinkModel;
import com.kazakago.cleanarchitecture.domain.model.weather.LocationModel;
import com.kazakago.cleanarchitecture.domain.model.weather.WeatherModel;

import java.util.List;

import io.reactivex.Observable;

/**
 * Created by tamura_k on 2016/12/08.
 */
public class WeatherMapper {

    public static WeatherModel map(@Nullable WeatherEntity entity) {
        if (entity != null) {
            WeatherModel weatherModel = new WeatherModel();
            weatherModel.setCityId(entity.getCityId());
            if (entity.getLocation() != null) {
                LocationModel locationModel = new LocationModel();
                locationModel.setArea(entity.getLocation().getArea());
                locationModel.setPrefecture(entity.getLocation().getPrefecture());
                locationModel.setCity(entity.getLocation().getCity());
                weatherModel.setLocation(locationModel);
            }
            weatherModel.setTitle(entity.getTitle());
            weatherModel.setLink(entity.getLink());
            weatherModel.setPublicTime(entity.getPublicTime());
            if (entity.getDescription() != null) {
                DescriptionModel descriptionModel = new DescriptionModel();
                descriptionModel.setText(entity.getDescription().getText());
                descriptionModel.setPublicTime(entity.getDescription().getPublicTime());
                weatherModel.setDescription(descriptionModel);
            }
            if (entity.getForecasts() != null) {
                List<ForecastModel> forecastModels = Observable.fromIterable(entity.getForecasts())
                        .map(forecastEntity -> {
                            ForecastModel forecastModel = new ForecastModel();
                            forecastModel.setTelop(forecastEntity.getTelop());
                            forecastModel.setDate(forecastEntity.getDate());
                            forecastModel.setDateLabel(forecastEntity.getDateLabel());
                            if (forecastEntity.getImage() != null) {
                                ImageModel imageModel = new ImageModel();
                                imageModel.setTitle(forecastEntity.getImage().getTitle());
                                imageModel.setHeight(forecastEntity.getImage().getHeight());
                                imageModel.setLink(forecastEntity.getImage().getLink());
                                imageModel.setUrl(forecastEntity.getImage().getUrl());
                                imageModel.setWidth(forecastEntity.getImage().getWidth());
                                forecastModel.setImage(imageModel);
                            }
                            return forecastModel;
                        })
                        .toList()
                        .blockingGet();
                weatherModel.setForecasts(forecastModels);
            }
            if (entity.getPinpointLocations() != null) {
                List<LinkModel> pinpointLocations = Observable.fromIterable(entity.getPinpointLocations())
                        .map(pinpointLocationEntity -> {
                            LinkModel pinpointLocationModel = new LinkModel();
                            pinpointLocationModel.setName(pinpointLocationEntity.getName());
                            pinpointLocationModel.setLink(pinpointLocationEntity.getLink());
                            return pinpointLocationModel;
                        })
                        .toList()
                        .blockingGet();
                weatherModel.setPinpointLocations(pinpointLocations);
            }
            if (entity.getCopyright() != null) {
                CopyrightModel copyrightModel = new CopyrightModel();
                copyrightModel.setTitle(entity.getCopyright().getTitle());
                copyrightModel.setLink(entity.getCopyright().getLink());
                if (entity.getCopyright().getImage() != null) {
                    ImageModel imageModel = new ImageModel();
                    imageModel.setTitle(entity.getCopyright().getImage().getTitle());
                    imageModel.setHeight(entity.getCopyright().getImage().getHeight());
                    imageModel.setLink(entity.getCopyright().getImage().getLink());
                    imageModel.setUrl(entity.getCopyright().getImage().getUrl());
                    imageModel.setWidth(entity.getCopyright().getImage().getWidth());
                    copyrightModel.setImage(imageModel);
                }
                if (entity.getCopyright().getProvider() != null) {
                    List<LinkModel> providers = Observable.fromIterable(entity.getCopyright().getProvider())
                            .map(providerEntity -> {
                                LinkModel providerModel = new LinkModel();
                                providerModel.setName(providerEntity.getName());
                                providerModel.setLink(providerEntity.getLink());
                                return providerModel;
                            })
                            .toList()
                            .blockingGet();
                    copyrightModel.setProvider(providers);
                }
                weatherModel.setCopyright(copyrightModel);
            }
            return weatherModel;
        } else {
            return null;
        }
    }

}

File: app/src/main/java/com/kazakago/cleanarchitecture/data/mapper/WeatherMapper.kt
package com.kazakago.cleanarchitecture.data.mapper

import com.kazakago.cleanarchitecture.data.entity.weather.*
import com.kazakago.cleanarchitecture.domain.model.weather.*
import io.realm.RealmList

/**
 * Created by tamura_k on 2016/12/08.
 */
object WeatherMapper {

    fun execute(entity: WeatherEntity): WeatherModel {
        val weather = WeatherModel()
        weather.cityId = entity.cityId
        weather.location = entity.location?.let {
            val location = LocationModel()
            location.area = it.area
            location.prefecture = it.prefecture
            location.city = it.city
            location
        }
        weather.title = entity.title
        weather.link = entity.link
        weather.publicTime = entity.publicTime
        weather.description = entity.description?.let {
            val description = DescriptionModel()
            description.text = it.text
            description.publicTime = it.publicTime
            description
        }
        weather.forecasts = entity.forecasts.map {
            val forecast = ForecastModel()
            forecast.telop = it.telop
            forecast.date = it.date
            forecast.dateLabel = it.dateLabel
            forecast.image = it.image?.let {
                val image = ImageModel()
                image.title = it.title
                image.height = it.height
                image.link = it.link
                image.url = it.url
                image.width = it.width
                image
            }
            forecast
        }
        weather.pinpointLocations = entity.pinpointLocations.map {
            val pinpointLocation = LinkModel()
            pinpointLocation.name = it.name
            pinpointLocation.link = it.link
            pinpointLocation
        }
        weather.copyright = entity.copyright?.let {
            val copyright = CopyrightModel()
            copyright.title = it.title
            copyright.link = it.link
            copyright.image = it.image?.let {
                val image = ImageModel()
                image.title = it.title
                image.height = it.height
                image.link = it.link
                image.url = it.url
                image.width = it.width
                image
            }
            copyright.provider = it.provider.map {
                val provider = LinkModel()
                provider.name = it.name
                provider.link = it.link
                provider
            }
            copyright
        }
        return weather
    }

    fun execute(model: WeatherModel): WeatherEntity {
        val weather = WeatherEntity()
        weather.cityId = model.cityId
        weather.location = model.location?.let {
            val location = LocationEntity()
            location.area = it.area
            location.prefecture = it.prefecture
            location.city = it.city
            location
        }
        weather.title = model.title
        weather.link = model.link
        weather.publicTime = model.publicTime
        weather.description = model.description?.let {
            val description = DescriptionEntity()
            description.text = it.text
            description.publicTime = it.publicTime
            description
        }
        val forecastRealmList = RealmList<ForecastEntity>()
        model.forecasts.forEach {
            val forecast = ForecastEntity()
            forecast.telop = it.telop
            forecast.date = it.date
            forecast.dateLabel = it.dateLabel
            forecast.image = it.image?.let {
                val image = ImageEntity()
                image.title = it.title
                image.height = it.height
                image.link = it.link
                image.url = it.url
                image.width = it.width
                image
            }
            forecastRealmList.add(forecast)
        }
        weather.forecasts = forecastRealmList
        val pinpointLocationsRealmList = RealmList<LinkEntity>()
        model.pinpointLocations.forEach {
            val pinpointLocation = LinkEntity()
            pinpointLocation.name = it.name
            pinpointLocation.link = it.link
            pinpointLocationsRealmList.add(pinpointLocation)
        }
        weather.pinpointLocations = pinpointLocationsRealmList
        weather.copyright = model.copyright?.let {
            val copyright = CopyrightEntity()
            copyright.title = it.title
            copyright.link = it.link
            copyright.image = it.image?.let {
                val image = ImageEntity()
                image.title = it.title
                image.height = it.height
                image.link = it.link
                image.url = it.url
                image.width = it.width
                image
            }
            val providerRealmList = RealmList<LinkEntity>()
            it.provider.forEach {
                val provider = LinkEntity()
                provider.name = it.name
                provider.link = it.link
                providerRealmList.add(provider)
            }
            copyright.provider = providerRealmList
            copyright
        }
        return weather
    }
}
